### DISCUSSION TITLE
The future of Atmos Weather: v3.0.0

### DISCUSSION THREAD

**Discussion by atticuscornett** [atticuscornett/AtmosWeather#32] [Announcements]
# The future of Atmos Weather: v3.0.0

## Development of Atmos Weather v3.0.0 is underway - and big changes are coming.

Atmos Weather v3.0.0 has been under development for over a month now. 3.0 introduces massive changes to the underlying architecture of Atmos Weather that should greatly benefit the codebase's longevity and maintainability. Atmos Weather was originally developed on a deadline for the Congressional App Challenge. As such, many bad decisions were made in the interest of meeting said deadline which made the codebase a spaghetti-filled nightmare to work with. Some of these issues were fixed with v1.1 and v2.0, but the primary issue remained: **Atmos Weather was written without any UI framework or components.** This meant a lot of inefficient weird DOM shenanigans and building HTML with string concatenation. (Yikes!) 3.0 is a total codebase overhaul that rewrites everything in Svelte and updates Capacitor to v6. This should make implementing new features in the future much faster and make contributing easier for other developers.

***TL;DR - Atmos Weather v3.0.0 is a near-total rewrite in Svelte and deals with years of tech debt.***

## 3.0 will introduce a new level of customizability.

Atmos Weather has always had customizability at its core. From allowing an unlimited number of locations to granular control of alerts by location, Atmos Weather has always been able to adapt to user needs. 3.0 will take this idea to a whole new level with the widget system.  Location pages will soon be comprised of customizable data widgets. Users will be able to choose what data to display for each location, and in the future, I plan to introduce niche weather widgets for aviation and other uncommon weather data. Don't want to customize? Atmos Weather will still have sensible defaults that most users won't have to edit at all.

***TL;DR - Atmos Weather v3.0.0 will take user customizability to the next level while still maintaining its ease of use.***

## 3.0 is a big undertaking.

If it wasn't already apparent, Atmos Weather 3.0 is a significant undertaking. It is planned to simultaneously be a total rewrite of the app and a major feature update. As such, this update is being completed in stages, and there will be many pre-releases. Stage 1 was just completed. It is the total rewrite of current features with Svelte. You can try it out right now by downloading `v3.0.0-prealpha.1`. Stage 2 will be focused on optimizing the new code, cleaning up old code remnants, and documenting everything. (After this stage is complete, v3.0.0 will be moved to the main branch.) Finally, stage 3 will be adding all the great new features planned. The goal is to have 3.0 finished in early January. With such a large update, there are bound to be many, many bugs. If you have a chance, please test out the pre-releases and test for bugs - it is very greatly appreciated.

***TL;DR - Atmos Weather v3.0.0 will have a lot of pre-releases. Bug testing is appreciated!***

### TL;DR TL;DR - Stay tuned, big things are coming!

**Comment by atticuscornett** [comment]
If it wasn't clear already, Atmos Weather v3.0 missed the January release goal. Unfortunately, right towards the release being ready, my testing found absolutely massive lag issues with the mobile version that are hard to track down. This meant that I could not complete the update before returning to college. I am hoping to work on Atmos Weather some this week, but I cannot give an ETA for v3.0.

**Reply by atticuscornett** [reply]
Alright, let's try this again. I have spent the past week tracking down the massive lag issues, and I am pleased to say that I have finally tracked down and fixed the sources of the problem. I am hoping to have the update out within the next month! (Of course, by saying that, I may have cursed more issues upon myself.) Better half a year late than never.

**Comment by atticuscornett** [comment]
The first beta for v3.0.0 is out now! Try it here: https://github.com/atticuscornett/AtmosWeather/releases/tag/v3.0.0-beta.1

**Comment by atticuscornett** [comment]
Atmos Weather v3.0.0 is here! Check it out here: https://github.com/atticuscornett/AtmosWeather/releases/tag/v3.0.0
